### PR TITLE
Nessus On Prem Capability

### PR DESCRIPTION
**Pull request by praetorian-rad**
Based on https://github.com/praetorian-inc/chariot-client/pull/2

### CODE REVIEW

File: pkg/asm/capabilities/nessus.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	} `json:"outputs"`
+}
+
+func NewNessus(job model.Job) model.Capability {
+	return &Nessus{Job: job, Asset: job.Target, XYZ: NewXYZ()}
+}
+
+func (task *Nessus) Match() bool {
+	ok := task.Job.Config["api_url"] != "" && task.Job.Config["access_key"] != "" && task.Job.Config["secret_key"] != ""
+	return task.Asset.Is("cloudflare") && ok
```
Comment (privateducky) [inline pkg/asm/capabilities/nessus.go:61]: cloudflare?

File: pkg/asm/capabilities/nessus.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+func (task *Nessus) Invoke() error {
+	url, key, secret := task.Job.Config["api_url"], task.Job.Config["access_key"], task.Job.Config["secret_key"]
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
```
Comment (privateducky) [inline pkg/asm/capabilities/nessus.go:69]: why InsecureSkipVerify?

File: pkg/asm/capabilities/nessus.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+			if err != nil {
+				continue
+			}
+
+			dns := host.Info.IP
+			if host.Info.FQDN != "" {
+				dns = host.Info.FQDN
+			}
+			asset := model.NewAsset(dns, host.Info.IP)
+			asset.Username = task.Job.Username
```
Comment (privateducky) [inline pkg/asm/capabilities/nessus.go:98]: why add the username? that should not be required.

File: pkg/asm/capabilities/nessus.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+			task.Job.Stream <- asset
+
+			for _, vuln := range host.Vulns {
+				if vuln.Severity == 0 { // Info
```
Comment (privateducky) [inline pkg/asm/capabilities/nessus.go:102]: good use of comment

File: pkg/asm/capabilities/nessus.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+					continue
+				}
+
+				plugin, err := request[Plugin](client, "GET", fmt.Sprintf("%s/scans/%d/hosts/%d/plugins/%d", url, scan.ID, info.HostID, vuln.PluginID), nil, headers...)
+				if err != nil {
+					continue
+				}
+
+				risk := model.NewRisk(asset, vuln.Name)
+				risk.Comment = plugin.Info.Attrs.Desc
```
Comment (privateducky) [inline pkg/asm/capabilities/nessus.go:112]: let's remove the comment. we're no longer using this to store persistent comments but instead apply comments to status change (history) events.

File: pkg/asm/capabilities/nessus.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+				risk.Status = model.TriageInfo
```
Comment (privateducky) [inline pkg/asm/capabilities/nessus.go:113]: this status is the default in risk.go, so we don't need to specify it.

File: pkg/asm/capabilities/nessus.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+				risk.Username = task.Job.Username
```
Comment (privateducky) [inline pkg/asm/capabilities/nessus.go:114]: username not required

File: pkg/asm/capabilities/nessus.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+				task.Job.Stream <- risk
+
+				var proof string
+				for _, output := range plugin.Outputs {
```
Comment (privateducky) [inline pkg/asm/capabilities/nessus.go:118]: I think this block of code handling proofs can be cleaned up and compressed a bit.

File: pkg/asm/capabilities/nessus.go (hunk 2 above)
Comment (privateducky) [reply pkg/asm/capabilities/nessus.go:69]: if it WAS required, given the impact to the other capabilities, I'd prefer us to create a "class" (struct) in xyz.go with the http.Client on it and let some callers (like this) overwrite the default with their own client.

however, before that, I'd want to know:
1. is this actually needed?
2. if yes, what is the impact of making this the default client in xyz.go?

**Review by privateducky — CHANGES_REQUESTED** [review]
reviewed

File: pkg/asm/capabilities/nessus.go (hunk 2 above)
Comment (praetorian-rad) [reply pkg/asm/capabilities/nessus.go:69]: this is needed as pretty much all on prem nessus setups don't have a valid certificate

**Review by privateducky — CHANGES_REQUESTED** [review]
1 primary topic (request function)

File: pkg/asm/capabilities/nessus.go (hunk 2 above)
Comment (UNC1739) [reply pkg/asm/capabilities/nessus.go:69]: I think this is more unique to the Nessus capability. Most companies with their on-premises appliances aren't using valid TLS certificates internally. Often it's going to be a self-signed certificate.

**Comment by praetorian-rad** [comment]
https://github.com/praetorian-inc/chariot-client/pull/23

**Pull request closed without merging**